File: frontend/src/api/queries/feature-lab.ts
/**
 * Feature Lab — API types, query keys, and fetchers for feature importance and stability analysis.
 *
 * NOTE: The fetchers normalise backend response shapes into the types below,
 * so component code can rely on a stable contract. The `Raw*` interfaces below
 * mirror the actual FastAPI response shapes from
 * `api/routers/forecasting/feature_lab.py`.
 */

import { fetchJson } from "./request";

// --- Types (frontend contract) ---

export interface FeatureImportanceRow {
  feature: string;
  shap_value: number;
  category: string;
  rank: number;
}

export interface FeatureStabilityRow {
  feature: string;
  mean_rank: number;
  rank_std: number;
  stability: "stable" | "moderate" | "unstable";
  n_folds: number;
}

export interface FeatureCorrelationCell {
  feature_a: string;
  feature_b: string;
  correlation: number;
}

export interface ClusterFeatureImportance {
  cluster: number;
  features: Array<{ feature: string; shap_value: number; rank: number }>;
}

export interface FeatureCategory {
  category: string;
  color: string;
  count: number;
}

// --- Raw wire shapes (mirror api/routers/forecasting/feature_lab.py) ---

/** One feature entry from GET /feature-lab/importance. */
interface RawImportanceFeature {
  name: string;
  mean_abs_shap: number;
  rank: number;
  selected_count: number;
  n_timeframes: number;
  category: string;
}

interface RawImportanceResponse {
  available: boolean;
  model_id: string;
  features: RawImportanceFeature[];
  total_features: number;
  selected_features: number;
}

/** One feature entry from GET /feature-lab/stability. */
interface RawStabilityFeature {
  name: string;
  ranks_by_timeframe: number[];
  mean_rank: number;
  rank_std: number;
  stability: "high" | "medium" | "unstable";
  min_rank: number;
  max_rank: number;
}

interface RawStabilityResponse {
  available: boolean;
  features: RawStabilityFeature[];
}

/** GET /feature-lab/correlation — features list plus a square correlation matrix. */
interface RawCorrelationResponse {
  available: boolean;
  features: string[];
  matrix: number[][];
  high_correlation_pairs: Array<{
    feature_a: string;
    feature_b: string;
    correlation: number;
    recommendation: string;
  }>;
}

/** GET /feature-lab/per-cluster-importance — cluster ids are returned as strings. */
interface RawPerClusterResponse {
  available: boolean;
  clusters: string[];
  features: string[];
  importance_matrix: number[][];
  cluster_specific_features: Array<{
    cluster: string;
    top_unique_feature: string;
    note: string;
  }>;
  note?: string;
}

/** One category entry from GET /feature-lab/categories. */
interface RawCategory {
  name: string;
  features: string[];
  description: string;
  count: number;
}

interface RawCategoriesResponse {
  available: boolean;
  model_id: string;
  categories: RawCategory[];
  total_features: number;
}

// --- Query keys ---

export const featureLabKeys = {
  all: ["feature-lab"] as const,
  importance: (modelId?: string) => [...featureLabKeys.all, "importance", modelId ?? "default"] as const,
  stability: () => [...featureLabKeys.all, "stability"] as const,
  correlation: (topN?: number) => [...featureLabKeys.all, "correlation", topN ?? 20] as const,
  clusterImportance: (cluster: number) => [...featureLabKeys.all, "cluster-importance", cluster] as const,
  categories: () => [...featureLabKeys.all, "categories"] as const,
};

// --- Default category colours (used when API doesn't provide colour) ---

const DEFAULT_CATEGORY_COLORS: Record<string, string> = {
  lag: "#2563EB",
  rolling: "#0D9488",
  seasonal: "#D97706",
  calendar: "#0891B2",
  cluster: "#EC4899",
  external: "#84CC16",
  demand: "#f97316",
  profile: "#8B5CF6",
  derived: "#06B6D4",
  categorical: "#F43F5E",
  fourier: "#D97706",
  croston: "#10B981",
  cross_dfu: "#6366F1",
  other: "#64748B",
};

// --- Stability mapping (API uses "high"/"medium"/"unstable"; frontend uses "stable"/"moderate"/"unstable") ---

function mapStability(s: string): "stable" | "moderate" | "unstable" {
  const lower = s.toLowerCase();
  if (lower === "high" || lower === "stable") return "stable";
  if (lower === "medium" || lower === "moderate") return "moderate";
  return "unstable";
}

// --- Fetchers ---

export async function fetchFeatureImportance(
  modelId?: string,
): Promise<{ features: FeatureImportanceRow[]; model_id: string }> {
  const sp = new URLSearchParams();
  if (modelId) sp.set("model_id", modelId);
  const raw = await fetchJson<RawImportanceResponse>(
    `/feature-lab/importance${sp.toString() ? `?${sp}` : ""}`,
  );
  const features: FeatureImportanceRow[] = (raw.features ?? []).map((f) => ({
    feature: f.name,
    shap_value: f.mean_abs_shap,
    category: f.category,
    rank: f.rank,
  }));
  return { features, model_id: raw.model_id ?? modelId ?? "lgbm_cluster" };
}

export async function fetchFeatureStability(): Promise<{ features: FeatureStabilityRow[] }> {
  const raw = await fetchJson<RawStabilityResponse>("/feature-lab/stability");
  const features: FeatureStabilityRow[] = (raw.features ?? []).map((f) => ({
    feature: f.name,
    mean_rank: f.mean_rank,
    rank_std: f.rank_std,
    stability: mapStability(f.stability),
    n_folds: Array.isArray(f.ranks_by_timeframe) ? f.ranks_by_timeframe.length : 0,
  }));
  return { features };
}

export async function fetchFeatureCorrelation(
  topN = 20,
): Promise<{ top_n: number; features: string[]; cells: FeatureCorrelationCell[] }> {
  const raw = await fetchJson<RawCorrelationResponse>(`/feature-lab/correlation?top_n=${topN}`);
  const features: string[] = raw.features ?? [];

  // Backend returns a square correlation matrix; flatten the upper triangle to non-zero cells.
  const matrix = raw.matrix ?? [];
  const cells: FeatureCorrelationCell[] = [];
  for (let i = 0; i < features.length; i++) {
    for (let j = i + 1; j < features.length; j++) {
      const corr = matrix[i]?.[j] ?? 0;
      if (corr !== 0) {
        cells.push({ feature_a: features[i], feature_b: features[j], correlation: corr });
      }
    }
  }

  return { top_n: features.length, features, cells };
}

export async function fetchClusterFeatureImportance(
  cluster: number,
): Promise<ClusterFeatureImportance> {
  const raw = await fetchJson<RawPerClusterResponse>("/feature-lab/per-cluster-importance");

  // Backend returns {clusters, features, importance_matrix} with string cluster ids —
  // extract the row for the requested cluster.
  const clusterStr = String(cluster);
  const clusterIdx = (raw.clusters ?? []).indexOf(clusterStr);

  if (clusterIdx === -1 || !raw.importance_matrix?.[clusterIdx]) {
    return { cluster, features: [] };
  }

  const featureNames: string[] = raw.features ?? [];
  const importanceRow: number[] = raw.importance_matrix[clusterIdx];

  const features = featureNames
    .map((name, i) => ({
      feature: name,
      shap_value: importanceRow[i] ?? 0,
      rank: 0,
    }))
    .filter((f) => f.shap_value > 0)
    .sort((a, b) => b.shap_value - a.shap_value);

  features.forEach((f, i) => {
    f.rank = i + 1;
  });

  return { cluster, features };
}

export async function fetchFeatureCategories(): Promise<{ categories: FeatureCategory[] }> {
  const raw = await fetchJson<RawCategoriesResponse>("/feature-lab/categories");
  const categories: FeatureCategory[] = (raw.categories ?? []).map((c) => ({
    category: c.name,
    color: DEFAULT_CATEGORY_COLORS[c.name] ?? "#64748B",
    count: c.count,
  }));
  return { categories };
}

File: api/routers/forecasting/feature_lab.py
"""Feature Lab — interactive feature importance exploration for LGBM tuning UI.

Reads SHAP output files from ``data/backtest/{model_id}/shap/`` and serves
feature importance, stability, correlation, per-cluster breakdown, and
categorization endpoints.  All endpoints are read-only (file-based, no DB).
"""
from __future__ import annotations

import logging
import re
from pathlib import Path
from typing import Any

import numpy as np
import pandas as pd
from fastapi import APIRouter, Query
from fastapi.responses import Response as FastAPIResponse

from api.core import set_cache

logger = logging.getLogger(__name__)

router = APIRouter(tags=["feature-lab"])

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

DATA_ROOT = Path("data/backtest")

# Feature category definitions — maps prefix/name patterns to category metadata
CATEGORY_DEFINITIONS: list[dict[str, Any]] = [
    {
        "name": "lag",
        "pattern_prefixes": ["qty_lag_"],
        "exact_names": [],
        "description": "Historical demand lags",
    },
    {
        "name": "rolling",
        "pattern_prefixes": ["rolling_mean_", "rolling_std_"],
        "exact_names": [],
        "description": "Rolling statistics",
    },
    {
        "name": "calendar",
        "pattern_prefixes": [],
        "exact_names": [
            "month", "quarter", "month_sin", "month_cos",
            "is_quarter_end", "is_year_end", "days_in_month",
        ],
        "description": "Calendar/seasonal signals",
    },
    {
        "name": "derived",
        "pattern_prefixes": ["lag_ratio_"],
        "exact_names": [
            "mom_growth", "demand_accel", "volatility_ratio",
            "n_zero_last_6m",
        ],
        "description": "Computed demand dynamics",
    },
    {
        "name": "profile",
        "pattern_prefixes": [],
        "exact_names": [
            "cv_demand", "zero_demand_pct", "trend_slope_norm",
            "recency_ratio", "seasonal_amplitude", "adi",
            "mean_demand", "yoy_correlation",
        ],
        "description": "Static DFU characteristics",
    },
    {
        "name": "categorical",
        "pattern_prefixes": [],
        "exact_names": [
            "ml_cluster", "region", "brand", "abc_vol",
            "execution_lag", "total_lt", "case_weight",
            "item_proof", "bpc",
        ],
        "description": "Categorical attributes",
    },
    {
        "name": "fourier",
        "pattern_prefixes": ["fourier_sin_", "fourier_cos_"],
        "exact_names": [],
        "description": "Fourier seasonal harmonics",
    },
    {
        "name": "croston",
        "pattern_prefixes": ["croston_"],
        "exact_names": [],
        "description": "Intermittent demand signals",
    },
    {
        "name": "cross_dfu",
        "pattern_prefixes": ["cluster_"],
        "exact_names": [],
        "description": "Cross-DFU cluster aggregates",
    },
    {
        "name": "external",
        "pattern_prefixes": ["ext_fcst_"],
        "exact_names": [],
        "description": "External forecast signals",
    },
]


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _classify_feature(name: str) -> str:
    """Return the category name for a given feature, or 'other'."""
    for cat in CATEGORY_DEFINITIONS:
        if name in cat["exact_names"]:
            return cat["name"]
        for prefix in cat["pattern_prefixes"]:
            if name.startswith(prefix):
                return cat["name"]
    return "other"


_MODEL_ID_PATTERN = re.compile(r"^[a-zA-Z0-9_-]{1,80}$")


def _shap_dir(model_id: str) -> Path:
    if not _MODEL_ID_PATTERN.match(model_id):
        raise ValueError(f"Invalid model_id: {model_id!r}")
    resolved = (DATA_ROOT / model_id / "shap").resolve()
    if not str(resolved).startswith(str(DATA_ROOT.resolve())):
        raise ValueError(f"Invalid model_id: {model_id!r}")
    return resolved


def _read_summary(model_id: str) -> pd.DataFrame | None:
    """Read the shap_summary.csv for a model.  Returns None if missing."""
    path = _shap_dir(model_id) / "shap_summary.csv"
    if not path.exists():
        return None
    try:
        return pd.read_csv(path)
    except (pd.errors.EmptyDataError, pd.errors.ParserError) as exc:
        logger.warning("Failed to read %s: %s", path, exc)
        return None


def _read_timeframe_files(model_id: str) -> list[pd.DataFrame]:
    """Read all shap_timeframe_*.csv files for a model."""
    shap_dir = _shap_dir(model_id)
    if not shap_dir.exists():
        return []
    frames: list[pd.DataFrame] = []
    for path in sorted(shap_dir.glob("shap_timeframe_*.csv")):
        try:
            df = pd.read_csv(path)
            if not df.empty:
                frames.append(df)
        except (pd.errors.EmptyDataError, pd.errors.ParserError) as exc:
            logger.warning("Failed to read %s: %s", path, exc)
    return frames


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------


@router.get("/feature-lab/importance")
def get_feature_importance(
    response: FastAPIResponse,
    model_id: str = Query(default="lgbm_cluster", max_length=120),
):
    """Current feature importance from latest backtest SHAP summary."""
    set_cache(response, max_age=120)

    summary = _read_summary(model_id)
    if summary is None:
        return {
            "available": False,
            "model_id": model_id,
            "features": [],
            "total_features": 0,
            "selected_features": 0,
        }

    features: list[dict[str, Any]] = []
    for _, row in summary.iterrows():
        name = str(row["feature"])
        features.append({
            "name": name,
            "mean_abs_shap": float(row["mean_abs_shap_across_timeframes"]),
            "rank": int(row["mean_rank"]),
            "selected_count": int(row["selected_count"]),
            "n_timeframes": int(row["n_timeframes"]),
            "category": _classify_feature(name),
        })

    # Sort by mean_abs_shap descending (highest importance first)
    features.sort(key=lambda f: f["mean_abs_shap"], reverse=True)
    # Assign rank based on sorted position
    for i, feat in enumerate(features, start=1):
        feat["rank"] = i

    n_timeframes = int(summary["n_timeframes"].iloc[0]) if len(summary) > 0 else 0
    selected = sum(1 for f in features if f["selected_count"] == n_timeframes)

    return {
        "available": True,
        "model_id": model_id,
        "features": features,
        "total_features": len(features),
        "selected_features": selected,
    }


@router.get("/feature-lab/stability")
def get_feature_stability(
    response: FastAPIResponse,
    model_id: str = Query(default="lgbm_cluster", max_length=120),
):
    """Feature rank stability across backtest timeframes."""
    set_cache(response, max_age=120)

    frames = _read_timeframe_files(model_id)
    if not frames:
        return {"available": False, "features": []}

    # Only use pooled (cluster=="all") rows for rank stability
    pooled_frames: list[pd.DataFrame] = []
    for df in frames:
        if "cluster" in df.columns:
            pooled = df[df["cluster"] == "all"]
        else:
            pooled = df
        pooled_frames.append(pooled)

    # Build feature → list of ranks across timeframes
    rank_map: dict[str, list[int]] = {}
    for df in pooled_frames:
        for _, row in df.iterrows():
            name = str(row["feature"])
            rank = int(row["rank"])
            rank_map.setdefault(name, []).append(rank)

    features: list[dict[str, Any]] = []
    for name, ranks in sorted(rank_map.items()):
        arr = np.array(ranks, dtype=np.float64)
        mean_rank = float(np.mean(arr))
        rank_std = float(np.std(arr, ddof=1)) if len(arr) > 1 else 0.0

        if rank_std < 2:
            stability = "high"
        elif rank_std < 5:
            stability = "medium"
        else:
            stability = "unstable"

        features.append({
            "name": name,
            "ranks_by_timeframe": ranks,
            "mean_rank": round(mean_rank, 1),
            "rank_std": round(rank_std, 1),
            "stability": stability,
            "min_rank": int(np.min(arr)),
            "max_rank": int(np.max(arr)),
        })

    # Sort by mean_rank ascending (most important first)
    features.sort(key=lambda f: f["mean_rank"])

    return {"available": True, "features": features}


@router.get("/feature-lab/correlation")
def get_feature_correlation(
    response: FastAPIResponse,
    model_id: str = Query(default="lgbm_cluster", max_length=120),
    top_n: int = Query(default=20, ge=2, le=100),
):
    """Feature importance correlation matrix across timeframes.

    Computes pairwise Pearson correlation of feature SHAP importance profiles
    across backtest timeframes for the top-N most important features.
    """
    set_cache(response, max_age=300)

    frames = _read_timeframe_files(model_id)
    if not frames:
        return {
            "available": False,
            "features": [],
            "matrix": [],
            "high_correlation_pairs": [],
        }

    # Build a matrix: rows=features, cols=timeframes, values=mean_abs_shap
    pooled_frames: list[pd.DataFrame] = []
    for df in frames:
        if "cluster" in df.columns:
            pooled = df[df["cluster"] == "all"]
        else:
            pooled = df
        pooled_frames.append(pooled)

    # Identify top-N features by average SHAP across timeframes
    shap_sums: dict[str, float] = {}
    for df in pooled_frames:
        for _, row in df.iterrows():
            name = str(row["feature"])
            shap_sums[name] = shap_sums.get(name, 0.0) + float(row["mean_abs_shap"])

    top_features = sorted(shap_sums, key=shap_sums.get, reverse=True)[:top_n]  # type: ignore[arg-type]

    if len(top_features) < 2:
        return {
            "available": True,
            "features": top_features,
            "matrix": [[1.0]] if top_features else [],
            "high_correlation_pairs": [],
        }

    # Build importance profile matrix: feature × timeframe
    n_tf = len(pooled_frames)
    profile_data: dict[str, list[float]] = {f: [0.0] * n_tf for f in top_features}
    for tf_idx, df in enumerate(pooled_frames):
        for _, row in df.iterrows():
            name = str(row["feature"])
            if name in profile_data:
                profile_data[name][tf_idx] = float(row["mean_abs_shap"])

    profile_df = pd.DataFrame(profile_data, index=range(n_tf))
    corr_matrix = profile_df.corr(method="pearson")

    # Replace NaN with 0 (features with constant SHAP across all timeframes)
    corr_matrix = corr_matrix.fillna(0.0)

    feature_list = list(corr_matrix.columns)
    matrix = corr_matrix.values.tolist()

    # Identify high-correlation pairs (|r| > 0.9, exclude diagonal)
    high_pairs: list[dict[str, Any]] = []
    for i in range(len(feature_list)):
        for j in range(i + 1, len(feature_list)):
            r = corr_matrix.iloc[i, j]
            if abs(r) > 0.9:
                recommendation = (
                    "consider dropping one"
                    if abs(r) > 0.95
                    else "keep both - different signals"
                )
                high_pairs.append({
                    "feature_a": feature_list[i],
                    "feature_b": feature_list[j],
                    "correlation": round(float(r), 4),
                    "recommendation": recommendation,
                })

    high_pairs.sort(key=lambda p: abs(p["correlation"]), reverse=True)

    return {
        "available": True,
        "features": feature_list,
        "matrix": matrix,
        "high_correlation_pairs": high_pairs,
    }


@router.get("/feature-lab/per-cluster-importance")
def get_per_cluster_importance(
    response: FastAPIResponse,
    model_id: str = Query(default="lgbm_cluster", max_length=120),
):
    """Feature importance breakdown by cluster from per-timeframe SHAP files."""
    set_cache(response, max_age=120)

    frames = _read_timeframe_files(model_id)
    if not frames:
        return {
            "available": False,
            "clusters": [],
            "features": [],
            "importance_matrix": [],
            "cluster_specific_features": [],
        }

    # Collect per-cluster rows (exclude "all" pooled rows)
    cluster_rows: list[pd.DataFrame] = []
    for df in frames:
        if "cluster" not in df.columns:
            continue
        per_cluster = df[df["cluster"] != "all"]
        if not per_cluster.empty:
            cluster_rows.append(per_cluster)

    if not cluster_rows:
        return {
            "available": False,
            "clusters": [],
            "features": [],
            "importance_matrix": [],
            "cluster_specific_features": [],
            "note": "No per-cluster SHAP data found (model may use global strategy)",
        }

    combined = pd.concat(cluster_rows, ignore_index=True)

    # Average mean_abs_shap per (cluster, feature) across timeframes
    agg = combined.groupby(["cluster", "feature"])["mean_abs_shap"].mean().reset_index()
    clusters = sorted(agg["cluster"].unique().tolist())
    all_features = sorted(agg["feature"].unique().tolist())

    # Build importance matrix: rows=clusters, cols=features
    importance_matrix: list[list[float]] = []
    for cluster in clusters:
        cluster_data = agg[agg["cluster"] == cluster]
        feat_map = dict(zip(cluster_data["feature"], cluster_data["mean_abs_shap"]))
        row = [round(float(feat_map.get(f, 0.0)), 4) for f in all_features]
        importance_matrix.append(row)

    # Identify cluster-specific top features
    # For each cluster, find the feature with the highest relative importance
    # compared to its average importance across clusters
    cluster_specific: list[dict[str, str]] = []
    if len(clusters) > 1:
        for ci, cluster in enumerate(clusters):
            cluster_importances = np.array(importance_matrix[ci])
            # Mean importance across all clusters for each feature
            all_importances = np.array(importance_matrix)
            mean_across_clusters = all_importances.mean(axis=0)
            # Relative lift: how much more important in this cluster vs average
            with np.errstate(divide="ignore", invalid="ignore"):
                lift = np.where(
                    mean_across_clusters > 0,
                    cluster_importances / mean_across_clusters,
                    0.0,
                )
            best_idx = int(np.argmax(lift))
            if lift[best_idx] > 1.2:  # At least 20% above average
                cluster_specific.append({
                    "cluster": str(cluster),
                    "top_unique_feature": all_features[best_idx],
                    "note": (
                        f"{all_features[best_idx]} is {lift[best_idx]:.1f}x "
                        f"more important in cluster {cluster} than average"
                    ),
                })

    return {
        "available": True,
        "clusters": clusters,
        "features": all_features,
        "importance_matrix": importance_matrix,
        "cluster_specific_features": cluster_specific,
    }


@router.get("/feature-lab/categories")
def get_feature_categories(
    response: FastAPIResponse,
    model_id: str = Query(default="lgbm_cluster", max_length=120),
):
    """Feature categorization — groups features by type with descriptions."""
    set_cache(response, max_age=600)

    summary = _read_summary(model_id)

    # If summary available, use its feature list; otherwise return static defs
    if summary is not None:
        feature_names = summary["feature"].tolist()
    else:
        feature_names = []

    # Build categorized output
    categorized: dict[str, list[str]] = {}
    for name in feature_names:
        cat = _classify_feature(name)
        categorized.setdefault(cat, []).append(name)

    categories: list[dict[str, Any]] = []
    for cat_def in CATEGORY_DEFINITIONS:
        cat_name = cat_def["name"]
        cat_features = categorized.pop(cat_name, [])
        categories.append({
            "name": cat_name,
            "features": sorted(cat_features),
            "description": cat_def["description"],
            "count": len(cat_features),
        })

    # Add 'other' for unclassified features
    other_features = categorized.pop("other", [])
    if other_features:
        categories.append({
            "name": "other",
            "features": sorted(other_features),
            "description": "Uncategorized features",
            "count": len(other_features),
        })

    return {
        "available": summary is not None,
        "model_id": model_id,
        "categories": categories,
        "total_features": len(feature_names),
    }
